src/metabase/api/card.clj

```diff
@@ -458,8 +472,13 @@
    result_metadata        [:maybe qr/ResultsMetadata]
    cache_ttl              [:maybe ms/PositiveInt]
    collection_preview     [:maybe :boolean]}
-  (let [card-before-update (t2/hydrate (api/write-check Card id)
-                                       [:moderation_reviews :moderator_details])]
+  (check-card-type-and-dataset card-updates)
+  (let [card-before-update     (t2/hydrate (api/write-check Card id)
+                                           [:moderation_reviews :moderator_details])
+        is-model-after-update? (cond
+                                (some? type)    (= "model" type)
+                                (some? dataset) dataset
+                                :else           (card/model? card-before-update))]
```

Comment:
This also seems to duplicate `ensure-type-and-dataset-are-consistent` logic

Reply:
not sure how this is a duplication, this is meant to whether the card is a model after the update.

Reply:
This code contains the logic: dataset=true implies type=model. That logic is duplicated in `ensure-type-and-dataset-are-consistent`. Here's an example of how to dedupe it
```suggestion
        is-model-after-update? (if (and (nil? type) (nil? dataset))
                                 (card/model? card-before-update)
                                 (card/model? (card/ensure-type-and-dataset-are-consistent card-updates)))]
```